### PR TITLE
Fix errors when drawing empty image.

### PR DESCRIPTION
**Pull request by derekdreery**
# Implement empty images

Whilst making an image viewer, I wanted a default value for the image to display. What I found is that if I used the default constructor `ImageBuf::empty` I was hitting a panic.

Even worse than this, when I then made tests for empty images and tried out the CI, I found that on macos it led to a segmentation fault. This would always be caught, since I think that pointers that don't point anywhere (to zero-sized types) are set to point so some rediculous memory address that tells the operating system/processor to issue a memory page fault, but still it's UB which in the Rust world is *bad*. I raised an [issue with core-foundations-rs](https://github.com/servo/core-foundation-rs/issues/431) where they said the would do an `assert` to check for empty `[u8]`s, but still that would be a panic in our code, like Windows and Cairo.

This patch adds code to handle the case where an empty image is both created and drawn, and essentially does nothing in both cases. I argue that this is expected behaviour, and that the empty image is a useful placeholder where you have an image that is optionally present (you avoid needing to use things like the `Either` widget). An alternative would be to check the image dimensions and panic when they are incorrect, but I think this is a less pleasant user experience. (If you disagree please say so on the comments below)

### CODE REVIEW

**Comment by derekdreery** [comment]
I'm not sure how to test this other than wait to see if it fixes the problems in druid. Suggestions welcome!

**Comment by cmyr** [comment]
@derekdreery a good way to test would be to add a test in `piet-common/tests`; these tests are run against the three major backends, so we can ensure that behaviour is consistent everywhere; your test can use the [`#[should_panic]` attribute](https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic).

**Comment by derekdreery** [comment]
I added a test, and it showed that actually I had not fixed the problem (which I now have). That's why tests are good! :)

**Comment by derekdreery** [comment]
@cmyr interesting! It seems that an empty image segfaults on Mac, and also probably on Windows (I'm not so up to speed on windows memory access violations).

**Comment by derekdreery** [comment]
I'm trying to write fixes for mac/win, but I am relying on CI to test my solutions for me.

File: piet-cairo/src/lib.rs
Comment (cmyr) [inline piet-cairo/src/lib.rs:336]: Is there no `Rect::is_empty`? Maybe just `src_rect().area() == 0.0`?

Comment (derekdreery) [reply piet-cairo/src/lib.rs:336]: Hmm, I could add a `Rect::is_empty`. The `area` solution will work also. I'll have a look at this after I've stopped windows/mac crashing.

**Comment by derekdreery** [comment]
@cmyr would you mind taking over this PR? There is a segfault in the macOS code when drawing an empty image. I can't debug it as I don't have access to a macos computer, and using the CI is too slow.

**Comment by cmyr** [comment]
@derekdreery if you would like to close this PR and then push your branch to `origin` and open a new PR (so I an easily add commits), I would be happy to fix the macOS stuff!

**Comment by derekdreery** [comment]
Please don't merge - I'm just experimenting using the CI.

**Comment by derekdreery** [comment]
Right @cmyr I think this is ready for review! :grinning: 

I realised in the end that I could use `cargo clippy` on mac if I added the darwin target, which meant that I only had to use the CI after I know it would compile.

I'll update the first comment with the final details and some discussion about whether we want this.

File: piet-direct2d/src/lib.rs
Comment (raphlinus) [inline piet-direct2d/src/lib.rs:548]: I think `dst_rect.is_empty()` might be clearer here.

File: piet-coregraphics/src/lib.rs
Comment (raphlinus) [inline piet-coregraphics/src/lib.rs:366]: Out of scope for this PR, which is fine, but this TODO bothers me. There should be a common impl with this and `draw_image` that takes care of interpolation mode.

**Review by raphlinus — APPROVED** [review]
Looks good, thanks. Improving robustness is important, and any way for user code to cause a segfault should be a priority.

File: piet-direct2d/src/lib.rs
Comment (derekdreery) [reply piet-direct2d/src/lib.rs:548]: Sorry I missed this, will fix before merge, and I'll have a quick look to make sure I didn't miss any more.

**Pull request merged**
